pkg/query/aggregation_label_rewriter.go

```diff
@@ -0,0 +1,137 @@
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+const (
+	aggregationLabelName            = "__rollup__"
+	aggregatedMetricNameGlobPattern = "{*:aggr,*:sum,*:count,*:avg,*:min,*:max}"
+)
+
+type AggregationLabelRewriter struct {
+	logger  log.Logger
+	metrics *aggregationLabelRewriterMetrics
+
+	enabled           bool
+	desiredLabelValue string
+
+	aggregatedMetricNameGlob glob.Glob
+}
+
+type aggregationLabelRewriterMetrics struct {
+	aggregationLabelRewriteSkippedCount    *prometheus.CounterVec
+	aggregationLabelRewrittenCount         *prometheus.CounterVec
+	aggregationLabelAddedCount             prometheus.Counter
+	aggregationLabelRewriterRuntimeSeconds *prometheus.HistogramVec
+}
+
+func newAggregationLabelRewriterMetrics(reg prometheus.Registerer, desiredLabelValue string) *aggregationLabelRewriterMetrics {
+	m := &aggregationLabelRewriterMetrics{}
+
+	m.aggregationLabelRewriteSkippedCount = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+		Name: "skipped_total",
+		Help: "Total number of times aggregation-label-rewriter was enabled but skipped",
+	}, []string{"reason"})
+	m.aggregationLabelRewrittenCount = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+		Name:        "label_rewritten_total",
+		Help:        "Total number of times aggregation-label-rewriter rewrote the aggregation label",
+		ConstLabels: prometheus.Labels{"new_value": desiredLabelValue},
+	}, []string{"old_value"})
+	m.aggregationLabelAddedCount = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "label_added_total",
+		Help:        "Total number of times aggregation-label-rewriter added the aggregation label",
+		ConstLabels: prometheus.Labels{"new_value": desiredLabelValue},
+	})
+	m.aggregationLabelRewriterRuntimeSeconds = promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+		Name: "total_runtime_seconds",
+		Help: "Runtime of aggregation-label-rewriter in seconds",
+	}, []string{"is_rewritten"})
+
+	return m
+}
+
+func NewNopAggregationLabelRewriter() *AggregationLabelRewriter {
+	return &AggregationLabelRewriter{
+		enabled: false,
+	}
+}
+
+func NewAggregationLabelRewriter(logger log.Logger, reg prometheus.Registerer, desiredLabelValue string) *AggregationLabelRewriter {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	g := glob.MustCompile(aggregatedMetricNameGlobPattern)
+	return &AggregationLabelRewriter{
+		enabled:                  desiredLabelValue != "",
+		logger:                   logger,
+		metrics:                  newAggregationLabelRewriterMetrics(reg, desiredLabelValue),
+		desiredLabelValue:        desiredLabelValue,
+		aggregatedMetricNameGlob: g,
+	}
+}
+
+func (a *AggregationLabelRewriter) Rewrite(ms []*labels.Matcher) []*labels.Matcher {
+	if a.enabled {
+		startOfRun := time.Now()
+		needsRewrite := false
+		skipReason := "no-name-matcher"
+		var aggregationLabelMatcher *labels.Matcher
+		aggregationLabelIndex := -1
+		for i := 0; i < len(ms); i++ {
```

Comment:
seems a bug here since you append new items to `ms` and this for loop would never end